modules/catalog/src/main/java/org/apache/ignite/internal/catalog/ClockWaiter.java

```diff
@@ -0,0 +1,157 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.catalog;
+
+import static java.util.concurrent.CompletableFuture.failedFuture;
+
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import org.apache.ignite.internal.hlc.ClockUpdateListener;
+import org.apache.ignite.internal.hlc.HybridClock;
+import org.apache.ignite.internal.hlc.HybridTimestamp;
+import org.apache.ignite.internal.logger.IgniteLogger;
+import org.apache.ignite.internal.logger.Loggers;
+import org.apache.ignite.internal.manager.IgniteComponent;
+import org.apache.ignite.internal.thread.NamedThreadFactory;
+import org.apache.ignite.internal.util.IgniteSpinBusyLock;
+import org.apache.ignite.internal.util.PendingComparableValuesTracker;
+import org.apache.ignite.internal.util.TrackerClosedException;
+import org.apache.ignite.lang.NodeStoppingException;
+
+/**
+ * Allows to wait for the supplied clock to reach a required timesdtamp. It only uses the clock itself,
+ * no SafeTime mechanisms are involved.
+ */
+public class ClockWaiter implements IgniteComponent {
+    private static final IgniteLogger LOG = Loggers.forClass(ClockWaiter.class);
+
+    private final String nodeName;
+    private final HybridClock clock;
+
+    private final IgniteSpinBusyLock busyLock = new IgniteSpinBusyLock();
+
+    private final AtomicBoolean stopGuard = new AtomicBoolean(false);
+
+    private final PendingComparableValuesTracker<HybridTimestamp, Void> nowTracker = new PendingComparableValuesTracker<>(
+            new HybridTimestamp(1, 0)
+    );
+
+    private final ClockUpdateListener updateListener = this::onUpdate;
+
+    private volatile ScheduledExecutorService scheduler;
+
+    public ClockWaiter(String nodeName, HybridClock clock) {
+        this.nodeName = nodeName;
+        this.clock = clock;
+    }
+
+    @Override
+    public void start() {
+        clock.addUpdateListener(updateListener);
+
+        scheduler = Executors.newSingleThreadScheduledExecutor(new NamedThreadFactory(nodeName + "-clock-waiter", LOG));
+    }
+
+    @Override
+    public void stop() throws Exception {
+        if (!stopGuard.compareAndSet(false, true)) {
+            return;
+        }
+
+        busyLock.block();
+
+        clock.removeUpdateListener(updateListener);
+
+        nowTracker.close();
+
+        scheduler.shutdownNow();
+        scheduler.awaitTermination(10, TimeUnit.SECONDS);
+    }
+
+    private void onUpdate(HybridTimestamp newTs) {
+        nowTracker.update(newTs, null);
```

Comment:
Should it be wrapped in the busy lock?

Reply:
The busy lock does not allow the clients of this class to submit any work to it. Here, it's handling of some work that was submitted earlier. It doesn't seem problematic to allow some of this work be finished. Or there is some problem here?

Reply:
Since the clock comes from the "outside world", it can still be used after this component has been stopped, therefore this method can be triggered. Is that ok?